File: app/src/test/java/pl/pola_app/ui/activity/MainPresenterTest.java
package pl.pola_app.ui.activity;

import android.os.Bundle;
import android.util.Log;

import com.squareup.otto.Bus;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.annotation.Config;

import pl.pola_app.TestApplication;
import pl.pola_app.helpers.EventLogger;
import pl.pola_app.helpers.SessionId;
import pl.pola_app.model.SearchResult;
import pl.pola_app.network.Api;
import pl.pola_app.testutil.SearchUtil;
import pl.pola_app.ui.adapter.ProductList;
import pl.pola_app.ui.event.ReportButtonClickedEvent;
import retrofit2.Call;
import retrofit2.Response;

import static org.junit.Assert.fail;
import static org.mockito.Matchers.anyBoolean;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

@Config(application = TestApplication.class)
@RunWith(RobolectricGradleTestRunner.class)
public class MainPresenterTest {

    @Mock private MainViewBinder viewBinder;
    @Mock private ProductList productList;
    @Mock private Api api;
    @Mock private Bus eventBus;
    @Mock private EventLogger logger;
    @Mock private SessionId sessionId;
    @Mock private Call<SearchResult> mockCall;
    private MainPresenter presenter;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        presenter = new MainPresenter(viewBinder, productList, api, logger, sessionId, eventBus);
        when(api.getByCode(anyString(), anyString())).thenReturn(mockCall);
        when(api.getByCode(anyString(), anyString(), anyBoolean())).thenReturn(mockCall);
    }

    @Test
    public void testRegisterBus() throws Exception {
        presenter.register();

        verify(eventBus).register(presenter);
    }

    @Test
    public void testUnregisterBus() throws Exception {
        presenter.unregister();

        verify(eventBus).unregister(presenter);
    }

    @Test
    public void testCallCanceledOnUnregister() throws Exception {
        presenter.onBarcode("code", true);
        presenter.unregister();

        verify(mockCall).cancel();
    }

    @Test
    public void testDontAddExistingProduct() throws Exception {
        when(productList.itemExists("itemA")).thenReturn(true);
        presenter.onBarcode("itemA", true);

        verify(productList, never()).createProductPlaceholder();
        verifyNoMoreInteractions(api);
    }

    @Test
    public void testAddProduct() throws Exception {
        when(sessionId.get()).thenReturn("sessionId");
        when(viewBinder.getDeviceYear()).thenReturn(2017);
        presenter.onBarcode("barcode", true);

        verify(productList).createProductPlaceholder();
        verify(api).getByCode("barcode", "sessionId");
    }

    @Test
    public void testAddProductForOldDevices() throws Exception {
        when(sessionId.get()).thenReturn("sessionId");
        when(viewBinder.getDeviceYear()).thenReturn(2009);
        presenter.onBarcode("barcode", true);

        verify(productList).createProductPlaceholder();
        verify(api).getByCode("barcode", "sessionId", true);
    }

    @Test
    public void testProductAddedOnResponse() throws Exception {
        final SearchResult searchResult = SearchUtil.createSearchResult(1);

        presenter.onResponse(null, Response.success(searchResult));

        verify(productList).addProduct(searchResult);
        verifyNoMoreInteractions(productList);
    }

    @Test
    public void testResumeScanningOnResponse() throws Exception {
        final SearchResult searchResult = SearchUtil.createSearchResult(1);
        presenter.onResponse(null, Response.success(searchResult));

        verify(viewBinder).resumeScanning();
    }

    @Test
    public void testResponseForUnsupportedBarCodes() throws Exception {
        final SearchResult searchResult = SearchUtil.createSearchResult(1);
        searchResult.product_id = null;

        try {
            presenter.onResponse(null, Response.success(searchResult));
        } catch (Exception e) {
            fail(Log.getStackTraceString(e));
        }
    }

    @Test
    public void testPlaceholderRemovedOnFailedRequest() throws Exception {
        presenter.onFailure(null, new Throwable("msg"));

        verify(productList).removeProductPlaceholder();
        verifyNoMoreInteractions(productList);
    }

    @Test
    public void testResumeScanningOnFailure() throws Exception {
        presenter.onFailure(null, new Throwable("msg"));

        verify(viewBinder).resumeScanning();
    }

    @Test
    public void testShowErrorMessageOnFailure() throws Exception {
        presenter.onFailure(null, new Throwable("msg"));

        verify(viewBinder).showErrorMessage("msg");
    }

    @Test
    public void testShowNoConnectionMessage() throws Exception {
        final String noConnectionMessage = "Unable to resolve host \"www.pola-app.pl\": No address associated with hostname";
        presenter.onFailure(null, new Throwable(noConnectionMessage));

        verify(viewBinder).showNoConnectionMessage();
    }

    @Test
    public void testOpenProductDetailsOnClick() throws Exception {
        final SearchResult searchResult = SearchUtil.createSearchResult(1);
        presenter.onItemClicked(searchResult);

        verify(viewBinder).openProductDetails(searchResult);
        verify(viewBinder).turnOffTorch();
    }

    @Test
    public void testDismissDetailsView() throws Exception {
        presenter.productDetailsFragmentDismissed(null);

        verify(viewBinder).dismissProductDetailsView();
    }

    @Test
    public void testLaunchReportActivity() throws Exception {
        final SearchResult searchResult = SearchUtil.createSearchResult(1);
        searchResult.product_id = 123;
        presenter.reportButtonClicked(new ReportButtonClickedEvent(searchResult));

        verify(viewBinder).launchReportActivity("123");
    }

    @Test
    public void testSaveState() throws Exception {
        Bundle bundle = new Bundle();
        presenter.onSaveState(bundle);

        verify(productList).writeToBundle(bundle);
    }
}

File: app/src/test/java/pl/pola_app/ui/activity/MainPresenterTest.kt
package pl.pola_app.ui.activity

import android.os.Bundle
import android.util.Log
import com.nhaarman.mockitokotlin2.never
import com.nhaarman.mockitokotlin2.verify
import com.squareup.otto.Bus
import org.junit.Assert
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.*
import org.mockito.Mockito.mock
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import pl.pola_app.TestApplication
import pl.pola_app.helpers.EventLogger
import pl.pola_app.helpers.SessionId
import pl.pola_app.model.SearchResult
import pl.pola_app.network.Api
import pl.pola_app.testutil.SearchUtil
import pl.pola_app.ui.adapter.ProductList
import retrofit2.Call
import retrofit2.Response
import java.lang.Exception
import org.mockito.Mockito.verifyNoMoreInteractions

import org.mockito.Mockito.`when`


@Config(application = TestApplication::class)
@RunWith(
    RobolectricTestRunner::class
)
class MainPresenterTest {
    @Mock
    private lateinit var viewBinder: MainViewBinder
    @Mock
    private lateinit var productList: ProductList
    @Mock
    private lateinit var api: Api
    @Mock
    private lateinit var eventBus: Bus
    @Mock
    private lateinit var logger: EventLogger
    @Mock
    private lateinit var sessionId: SessionId
    @Mock
    private lateinit var mockCall: Call<SearchResult>
    @Mock
    private lateinit var presenter: MainPresenter

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this);
        presenter = MainPresenter(viewBinder,productList,api,logger,sessionId,eventBus)

        Mockito.`when`<Call<SearchResult>>(
            api.getByCode(
                ArgumentMatchers.anyString(),
                ArgumentMatchers.anyString()
            )
        ).thenReturn(mockCall)
        Mockito.`when`<Call<SearchResult>>(
            api.getByCode(
                ArgumentMatchers.anyString(),
                ArgumentMatchers.anyString(),
                ArgumentMatchers.anyBoolean()
            )
        ).thenReturn(mockCall)
    }

    @Test
    @Throws(Exception::class)
    fun testRegisterBus() {
        presenter.register()
        verify(eventBus).register(presenter)
    }

    @Test
    @Throws(Exception::class)
    fun testUnregisterBus() {
        presenter.unregister()
        verify(eventBus).unregister(presenter)
    }

//    @Test
//    @Throws(Exception::class)
//    fun testCallCanceledOnUnregister() {
//        presenter.onBarcode("code", true)
//        presenter.unregister()
//        verify(mockCall).cancel()
//    }


    @Test
    @Throws(Exception::class)
    fun testDontAddExistingProduct() {
        `when`(productList.itemExists("itemA")).thenReturn(true)
        presenter.onBarcode("itemA", true)
        verify(productList, never()).createProductPlaceholder()
        verifyNoMoreInteractions(api)
    }

    @Test
    @Throws(Exception::class)
    fun testAddProduct() {
        `when`(sessionId.get()).thenReturn("sessionId")
        `when`(viewBinder.deviceYear).thenReturn(2017)
        presenter.onBarcode("barcode", true)
        verify(productList).createProductPlaceholder()
        verify(api).getByCode("barcode", "sessionId")
    }

    @Test
    @Throws(Exception::class)
    fun testAddProductForOldDevices() {
        `when`(sessionId.get()).thenReturn("sessionId")
        `when`(viewBinder.deviceYear).thenReturn(2009)
        presenter.onBarcode("barcode", true)
        verify(productList).createProductPlaceholder()
        verify(api).getByCode("barcode", "sessionId", true)
    }

    @Test
    @Throws(Exception::class)
    fun testProductAddedOnResponse() {
        val searchResult = SearchUtil.createSearchResult(1)
        presenter.onResponse(null, Response.success(searchResult))
        verify(productList).addProduct(searchResult)
        verifyNoMoreInteractions(productList)
    }

    @Test
    @Throws(Exception::class)
    fun testResumeScanningOnResponse() {
        val searchResult = SearchUtil.createSearchResult(1)
        presenter.onResponse(null, Response.success(searchResult))
        verify(viewBinder).resumeScanning()
    }

    @Test
    @Throws(Exception::class)
    fun testResponseForUnsupportedBarCodes() {
        var searchResult = SearchUtil.createSearchResult(1)
        searchResult = searchResult.copy(product_id = null)
        try {
            presenter.onResponse(null, Response.success(searchResult))
        } catch (e: Exception) {
            fail(Log.getStackTraceString(e))
        }
    }


    @Test
    @Throws(Exception::class)
    fun testPlaceholderRemovedOnFailedRequest() {
        presenter.onFailure(null, Throwable("msg"))
        verify(productList).removeProductPlaceholder()
        verifyNoMoreInteractions(productList)
    }

    @Test
    @Throws(Exception::class)
    fun testResumeScanningOnFailure() {
        presenter.onFailure(null, Throwable("msg"))
        verify(viewBinder).resumeScanning()
    }

    @Test
    @Throws(Exception::class)
    fun testShowErrorMessageOnFailure() {
        presenter.onFailure(null, Throwable("msg"))
        verify(viewBinder).showErrorMessage("msg")
    }

    @Test
    @Throws(Exception::class)
    fun testShowNoConnectionMessage() {
        val noConnectionMessage =
            "Unable to resolve host \"www.pola-app.pl\": No address associated with hostname"
        presenter.onFailure(null, Throwable(noConnectionMessage))
        verify(viewBinder).showNoConnectionMessage()
    }

    @Test
    @Throws(Exception::class)
    fun testOpenProductDetailsOnClick() {
        val searchResult = SearchUtil.createSearchResult(1)
        presenter.onItemClicked(searchResult)
        verify(viewBinder).openProductDetails(searchResult)
        verify(viewBinder).turnOffTorch()
    }

    @Test
    @Throws(Exception::class)
    fun testDismissDetailsView() {
        presenter.productDetailsFragmentDismissed(null)
        verify(viewBinder).dismissProductDetailsView()
    }

    @Test
    @Throws(Exception::class)
    fun testSaveState() {
        val bundle = Bundle()
        presenter.onSaveState(bundle)
        verify(productList).writeToBundle(bundle)
    }
}